core/node/da_dispatcher/src/da_dispatcher.rs

```diff
@@ -0,0 +1,203 @@
+                )
+            })?;
+            let dispatch_latency_duration = dispatch_latency.observe();
+
+            let sent_at =
+                NaiveDateTime::from_timestamp_millis(Utc::now().timestamp_millis()).unwrap();
+
+            let mut conn = self.pool.connection_tagged("da_dispatcher").await?;
+            conn.data_availability_dal()
+                .insert_l1_batch_da(
+                    batch.l1_batch_number,
+                    dispatch_response.blob_id.as_str(),
+                    sent_at,
+                )
+                .await?;
+            drop(conn);
+
+            METRICS
+                .last_dispatched_l1_batch
+                .set(batch.l1_batch_number.0 as usize);
+            METRICS.blob_size.observe(batch.pubdata.len());
+            tracing::info!(
+                "Dispatched a DA for batch_number: {}, pubdata_size: {}, dispatch_latency: {dispatch_latency_duration:?}",
+                batch.l1_batch_number,
+                batch.pubdata.len(),
+            );
+        }
+
+        Ok(())
+    }
+
+    /// Polls the data availability layer for inclusion data, and saves it in the database.
+    async fn poll_for_inclusion(&self) -> anyhow::Result<()> {
+        let mut conn = self.pool.connection_tagged("da_dispatcher").await?;
+        let blob_info = conn
+            .data_availability_dal()
+            .get_first_da_blob_awaiting_inclusion()
+            .await?;
+        drop(conn);
+
+        let Some(blob_info) = blob_info else {
+            return Ok(());
+        };
+
+        let inclusion_data = self
+            .client
+            .get_inclusion_data(blob_info.blob_id.as_str())
+            .await
+            .with_context(|| {
+                format!(
+                    "failed to get inclusion data for blob_id: {}, batch_number: {}",
+                    blob_info.blob_id, blob_info.l1_batch_number
+                )
+            })?;
+
+        let Some(inclusion_data) = inclusion_data else {
+            return Ok(());
+        };
+
+        let mut conn = self.pool.connection_tagged("da_dispatcher").await?;
+        conn.data_availability_dal()
+            .save_l1_batch_inclusion_data(
+                L1BatchNumber(blob_info.l1_batch_number.0),
+                inclusion_data.data.as_slice(),
+            )
+            .await?;
+        drop(conn);
+
+        let inclusion_latency = Utc::now().signed_duration_since(blob_info.sent_at);
+        if let Ok(latency) = inclusion_latency.to_std() {
+            METRICS.inclusion_latency.observe(latency);
+        }
+        METRICS
+            .last_included_l1_batch
+            .set(blob_info.l1_batch_number.0 as usize);
+
+        tracing::info!(
+            "Received an inclusion data for a batch_number: {}, inclusion_latency_seconds: {}",
+            blob_info.l1_batch_number,
+            inclusion_latency.num_seconds()
+        );
+
+        Ok(())
+    }
+}
+
+async fn retry<T, Fut, F>(
```

Comment:
This function doesn't seem to account for `stop_receiver`, meaning that the dispatcher may delay the graceful shutdown of the node.

Reply:
added a `stop_receiver`